### PR TITLE
Enable perfect forwarding with irept/exprt/typet constructors

### PR DESCRIPTION
**Pull request by tautschnig**
This should enable use (with performance benefit) of rvalue references in
higher-level APIs.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by peterschrammel — COMMENTED** [review]
> n/a My commit message includes data points confirming performance improvements (if claimed)

Isn't your commit message claiming exactly that? ;)

**Comment by tautschnig** [comment]
> Isn't your commit message claiming exactly that? ;)

No, I'm not yet claiming that: I'm working on a PR that will actually start making serious use of these functions, at which point I should likely do a proper evaluation.

**Comment by tautschnig** [comment]
Although I have a fix for the test failures I'm marking this do-not-merge as I'll add the corresponding changes to *typet/\*exprt/code\*t in a single branch to do performance benchmarking. (Also, the failing tests did confirm that the functions _are_ used.)

File: src/util/irep.cpp
Comment (kroening) [inline src/util/irep.cpp:219]: Does the above work? I'd be worried that the first std::move gets evaluated, and then second call doesn't have anything to move anymore.

Comment (tautschnig) [reply src/util/irep.cpp:219]: No, it did not. I've got an implementation that does, will push soon - just working on PRs that set up proper use of all this.

**Comment by tautschnig** [comment]
Rvalue constructors added. I will now start testing and benchmarking. Various preparatory changes have been factored out into other PRs, and there is a lot that could and should be done on top of this one.

File: src/util/irep.cpp
Comment (smowton) [inline src/util/irep.cpp]: How about just `s[name] = std::move(irep);`?

File: src/util/std_code.h
Comment (smowton) [inline src/util/std_code.h]: Style: spacing of `:` (here and below)

**Comment by smowton** [comment]
Paging @NathanJPhillips for C++ wizardry: If you have a function with overloads `f(const xyz &)` and `f(xyz &&)`, and a function that takes a few xyzs and applies `f` to each one, is there a good way to get optimal use of `f`? For example, in this PR we have something like

```
g(const xyz &a, const xyz &b, const xyz &c)
{
  f(a); f(b); f(c);
}

g(xyz &&a, xyz &&b, xyz &&c)
{
  f(std::move(a)); f(std::move(b)); f(std::move(c));
}
```

But, (a) it's annoying to duplicate every method like `g`, which here means all the constructors on `exprt` children, and (b) this doesn't catch the case where a, b and c are mixed lvalue and rvalue refs.

I'd like a case like
```
template<class A, class B, class C>
g(A &&a, B &&b, C &&c)
{
  f(std::forward(a)); f(std::forward(b)); f(std::forward(c));
}
```

But it seems odd to have A, B and C be that general when actually I only want to allow `const xyz &` or `xyz &&`. Is there a nice way to do this?

File: src/util/irep.h
Comment (martin-cs) [inline src/util/irep.h]: I have feelings on this subject and I feel compelled to share them with you.

1. Before this can be merged we *need* to add the assertion to the simplifier that checks idempotency.  By design it should hold but ... I've been known to bet people up to 50p that it doesn't in one or more obscure cases.

2. We should also think about the cache in the simplifier.

3. Crazy idea : rather than just a bit, we could make simplification just a bit more union-find and have a pointer to the canonical version.  This would allow fast simplification and deduplification / re-introduction of sharing.

4. This road starts to feel like it leads to canonical irept's.  I remember discussing this ages ago and coming to the conclusion that this was common in SMT solvers because their nodes were simple and unchanging while CPROVER really needed them to be able to gain annotations / fields.  If so this could be ... a thing.

**Comment by tautschnig** [comment]
After a lot of back&forth in this branch to play with several optimisations this PR now does what the $SUBJECT says. Profiling is still running to make sure I haven't messed  up, but reviews would be appreciated.

File: src/util/irep.cpp
Comment (tautschnig) [reply src/util/irep.cpp]: We also want to return a reference to the newly added `irept`.

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: f9bfdb0cb645d945e6358595a23c2ff9be7ccc2b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100651205
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: e224f00a5c1b0853a7dbbe7f018ea9c25225891f).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100654772

**Comment by tautschnig** [comment]
No obvious regressions in my profiling results thus far, but now it's also possible to identify some offenders (first candidate: `object_descriptor_exprt()`). I'll keep reviewing the profile and will add small changes to this PR as needed.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: d272452bf5c4b49becbd350aaa6ce12d630f1829).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100672863

**Comment by kroening** [comment]
> Paging @NathanJPhillips for C++ wizardry: If you have a function with overloads `f(const xyz &)` and `f(xyz &&)`, and a function that takes a few xyzs and applies `f` to each one, is there a good way to get optimal use of `f`? For example, in this PR we have something like

For data structures that have an efficient move constructor it's hardly ever useful to have both the `const &` and the `&&` variant -- passing by value is the most efficient answer. If the thing being passed is stored, then use pass by value plus `std::move`, if it's just "inspected", then use "const &". In particular, this avoids the need to contemplate the cross-product of `const &` and `&&` when multiple arguments are given.

**Comment by kroening** [comment]
As example, the program
```
class data
{
public:
  data() { }
  data(const data &) { }
  data(data &&) { }
};

void func(data d1)
{
  data d2 = std::move(d1);
}

int main()
{
  func(data());
}
```
never actually calls that copy constructor.

**Comment by tautschnig** [comment]
Just for the record: @NathanJPhillips had shared his knowledge with me offline back in December, which is completely in line with what @kroening said above. (And profiling confirms that the compiler is doing what is expected of it.)

**Comment by tautschnig** [comment]
I'll keep benchmarking and adding minor nudges to this as I find them in the profiling data, but otherwise this is looking ok from a performance point of view. We should certainly start thinking about factories that construct expressions like `true_exprt` just once (I'll know in a few hours whether this is now a bit better, but the data from this morning suggested that we spend 200 seconds on 5B `true_exprt` constructions (in my current SV-COMP/ReachSafety-ECA set)).

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: d871239a5aae1703cb60a467b4fca8b0cf2da4da).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100949504
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: c58107c794bcf028538ea1a86f22b8fa906d4b9b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/100955074

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 77d73c1ece8d8831f9dcea68eb5bd253f638ff8d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/101066127

File: src/util/expr.h
Comment (kroening) [inline src/util/expr.h:69]: moving a `dstring` may well be slower than copying it.

File: src/util/irep.h
Comment (kroening) [inline src/util/irep.h:445]: Since this will copy eventually, simply have one variant, with just `irept`, and then move that?

File: src/util/irep.h
Comment (kroening) [inline src/util/irep.h:425]: One variant with just `irept` suffices.

**Comment by kroening** [comment]
You sometimes use the `.swap(other)` and sometimes the `= std::move(other)` pattern -- this will ultimately produce the same code, but it may nevertheless be nice to stick to one pattern. I suppose the `std::move` variant is the one that's more modern.

**Comment by tautschnig** [comment]
> You sometimes use the .swap(other) and sometimes the = std::move(other) pattern -- this will ultimately produce the same code, but it may nevertheless be nice to stick to one pattern. I suppose the std::move variant is the one that's more modern.

I used to have a commit that would clean this up across the codebase, but eventually got tired of maintaining it. I'll at least clean up those bits that affect this PR.

File: src/util/expr.h
Comment (tautschnig) [reply src/util/expr.h:69]: Will remove.

Comment (tautschnig) [reply src/util/expr.h:69]: Or actually: shouldn't we just define a move constructor for `dstringt` that actually is cheap? For the case that `irep_idt == std::string` we'd still want the `std::move`.

Comment (tautschnig) [reply src/util/expr.h:69]: I have added a commit that adds efficient move construction/assignment to `dstringt`.

File: src/util/irep.h
Comment (tautschnig) [reply src/util/irep.h:445]: Done.

File: src/util/irep.h
Comment (tautschnig) [reply src/util/irep.h:425]: Done.

**Review by allredj — COMMENTED** [review]
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 8cc4c5eabcc85eaec99065bf5972b39ce4292d11).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/103558660
Status will be re-evaluated on next push.
Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

  - the compatibility was already broken by an earlier merge.

File: src/util/expr.h
Comment (kroening) [reply src/util/expr.h:69]: I am still not entirely convinced that we need those `std::move(id)` -- we wouldn't do it for an integer, say.

File: src/util/ssa_expr.h
Comment (kroening) [inline src/util/ssa_expr.h:22]: I kind of liked that `symbol_exprt` now enforces that you set the identifier.

File: src/util/expr.h
Comment (tautschnig) [reply src/util/expr.h:69]: But we don't know whether `irep_idt` is a `std::string` (where it would seem worthwhile) or a `dstringt` (where it's certainly unnecessary). But also that isn't something I want this PR to get blocked on, I'm happy to follow whatever I'm advised to do.

File: src/util/std_code.h
Comment (kroening) [inline src/util/std_code.h:2091]: You could make the above `multi_ary_exprt(ID_arguments, std::move(_arguments))`

File: src/util/ssa_expr.h
Comment (tautschnig) [reply src/util/ssa_expr.h:22]: I don't think `symbol_exprt` actually does? This change here just removes an unnecessary call to `set_identifier`.

File: src/util/std_code.h
Comment (tautschnig) [reply src/util/std_code.h:2091]: Done.

**Review by allredj — COMMENTED** [review]
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 4b20ea0a1e72840e180180d1191460fe9bc1f1c3).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/104425235
Status will be re-evaluated on next push.
Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

  - the compatibility was already broken by an earlier merge.

File: src/util/std_expr.h
Comment (kroening) [inline src/util/std_expr.h:3172]: Good catch

Comment (tautschnig) [reply src/util/std_expr.h:3172]: ~Hmm, but then I should really add `as_const()` in there ;-)~ Edit: Done.

**Review by kroening — APPROVED** [review]

File: src/util/ssa_expr.h
Comment (kroening) [reply src/util/ssa_expr.h:22]: No, it doesn't, but I was hoping it would eventually.

**Review by allredj — COMMENTED** [review]
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 0520791d000f596d875001a97607f169441cccd1).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/104426210
Status will be re-evaluated on next push.
Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

  - the compatibility was already broken by an earlier merge.

File: src/util/ssa_expr.h
Comment (tautschnig) [reply src/util/ssa_expr.h:22]: We can certainly put the `irep_idt()` back if and when we do.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 83ed9a5acb93e1546166e0e0b754ea7dd572abfa).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/104448810

File: src/util/std_expr.h
Comment (kroening) [reply src/util/std_expr.h:3172]: Unfortunately, this doesn't do the job -- the mere access to `operands()` is enough to wreck the sharing.
This needs to be `as_const(*this).type().id()`.

Comment (tautschnig) [reply src/util/std_expr.h:3172]: My apologies for failing to take into account this remark right here, the fix is now in #4391.

**Pull request merged**
